pkg/observable/channel/observable.go

```diff
@@ -39,112 +43,126 @@ func NewObservable[V any](opts ...option[V]) (observable.Observable[V], chan<- V
-		activeObservers[idx] = toClose
-	}
-	// unlock before unsubscribing to avoid deadlock
-	obsvbl.observersMu.Unlock()
-
-	for _, observer := range activeObservers {
+// unsubscribeAll unsubscribes and removes all observers from the observable.
+func (obsvbl *channelObservable[V]) unsubscribeAll() {
+	// Copy currentObservers to avoid holding the lock while unsubscribing them.
+	// The observers at the time of locking, prior to copying, are the canonical
+	// set of observers which are unsubscribed.
+	// New or existing Observers may (un)subscribe while the observable is closing.
+	// Any such observers won't be isClosed but will also stop receiving notifications
+	// immediately (if they receive any at all).
+	currentObservers := obsvbl.copyObservers()
+	for _, observer := range currentObservers {
 		observer.Unsubscribe()
 	}
 
-	// clear observers
+	// Reset observers to an empty list. This purges any observers which might have
+	// subscribed while the observable was closing.
 	obsvbl.observersMu.Lock()
 	obsvbl.observers = []*channelObserver[V]{}
 	obsvbl.observersMu.Unlock()
 }
 
-// goProduce to the producer and notify observers when values are received. This
-// function is blocking and should be run in a goroutine.
-func (obsvbl *channelObservable[V]) goProduce(producer <-chan V) {
-	var observers []*channelObserver[V]
-	for notification := range producer {
-		for {
-			if !obsvbl.observersMu.TryRLock() {
-				continue
-			}
-			observers = make([]*channelObserver[V], len(obsvbl.observers))
-			for i, obsvr := range obsvbl.observers {
-				observers[i] = obsvr
-			}
-			obsvbl.observersMu.RUnlock()
-			break
-		}
-
-		// notify observers
-		for _, obsvr := range observers {
-			// CONSIDERATION: perhaps continue trying to avoid making this
+// goPublish to the publishCh and notify observers when values are received.
+// This function is blocking and should be run in a goroutine.
+func (obsvbl *channelObservable[V]) goPublish(publisher <-chan V) {
+	for notification := range publisher {
+		// Copy currentObservers to avoid holding the lock while notifying them.
+		// New or existing Observers may (un)subscribe while this notification
+		// is being fanned out.
+		// The observers at the time of locking, prior to copying, are the canonical
+		// set of observers which receive this notification.
+		currentObservers := obsvbl.copyObservers()
+		for _, obsvr := range currentObservers {
+			// TODO_CONSIDERATION: perhaps continue trying to avoid making this
 			// notification async as it would effectively use goroutines
-			// in memory as a buffer (with little control surface).
+			// in memory as a buffer (unbounded).
 			obsvr.notify(notification)
 		}
 	}
 
-	// Here we know that the producer has been closed, all observers should be closed as well
-	obsvbl.close()
+	// Here we know that the publisher channel has been closed.
+	// Unsubscribe all observers as they can no longer receive notifications.
+	obsvbl.unsubscribeAll()
```

Comment:
Moving previous convo here: https://github.com/pokt-network/poktroll/pull/31#discussion_r1364521938

Thoughts on doing a `if (len(obsvbl.observers) == 0) pain("Should have no observers after unsubscribing")`

Reply:
I don't think we want to do this. The observers list is explicitly being set to an empty slice at the end of `unsubscribeAll()`. If we wanted to check that the `obervers` slice is empty here, we would have to acquire a lock on the mutex. This raises some concerns for me with regards to false positives on the timing; although, this change doesn't seem to break tests. Regardless, this could be refactored by moving such a check directly into `unsubscribeAll()`. But then you end up with something that looks like this:

```go
// unsubscribeAll unsubscribes and removes all observers from the observable.
func (obsvbl *channelObservable[V]) unsubscribeAll() {
	...

	// Reset observers to an empty list. This purges any observers which might have
	// subscribed while the observable was closing.
	obsvbl.observersMu.Lock()
	obsvbl.observers = []*channelObserver[V]{}
	if (obsvbl.observers) > 0 {
		panic("impossible")
	}
	obsvbl.observersMu.Unlock()
}
```